working/333 - shared memory multithreading/proposal.md

```diff
@@ -0,0 +1,1582 @@
+synchronous invocation. There are obvious ways to address this:
+
+1. Create a variation of `NativeCallable.isolateLocal` which enters (and after
+   call leaves) the target isolate if the native caller is not already in the
+   target isolate.
+
+2. Create `NativeCallable.onTemporaryIsolate` which spawns (and after call
+   destroys) a temporary isolate to handle the call.
+
+Neither of these are truly satisfactory:
+
+- Allowing `isolateLocal` to enter target isolate means that it can block the
+  caller if the target isolate is busy doing _something_ (e.g. processing some
+  events) in parallel. This is unacceptable in situations when the caller is
+  latency sensitive e.g. audio thread or even just the main UI thread of an
+  application.
+- Using temporary isolate comes with a bunch of ergonomic problems as every
+  invocation is handled using a freshly created environment and no static state
+  is carried between invocations - which might surprise the developer. Sharing
+  mutable data between invocations requires storing it outside of Dart heap
+  using `dart:ffi`.
+
+> [!NOTE]
+>
+> This particular example might not look entirely convincing because it can be
+> reasonably well solved within confines of isolate model.
+>
+> When you look at an _isolate_ as _a bag of state guarded by a mutex_, you
+> eventually realize that this bag is simply way too big - it encompasses the
+> static state of the whole program - and this is what makes isolates unhandy to
+> use. The rule of thumb is that _coarse locks lead to scalability problems_.
+>
+> How do you solve it?
+>
+> _Spawn an isolate that does one specific small task_. In the context of this
+> example:
+>
+> - One isolate (_consumer_) is spawned to do nothing but synchronously handle
+>   `DataCallback` calls (using an extension of `isolateLocal` which enters and
+>   leaves isolate is required).
+> - Another isolate (_producer_) is responsible for generating the data which is
+>   fed to the audio-library. The data is allocated in the native heap and
+>   directly _shared_ with _consumer_.
+>
+> However, isolates don't facilitate this style of programming. They are too
+> _coarse_ - so it is easy to make a mistake, touch a static state you are not
+> supposed to touch, call a dependency which schedules an asynchronous task,
+> etc. Furthermore, you do still need a low overhead communication channel
+> between isolates. The shared memory is _still_ part of the solution here, even
+> though in this particular example we can manage with what `dart:ffi` allows
+> us. And that is, in my opinion, a pretty strong signal in favor of more shared
+> memory support in the language.
+
+Another variation of this problem occurs when trying to use Dart for sharing
+business logic and creating shared libraries. Imagine that `dart:ffi` provided a
+way to export static functions as C symbols:
+
+```dart
+// foo.dart
+import 'dart:ffi';
+
+@ffi.Export()
```

Comment:
Maybe add a link to https://github.com/dart-lang/sdk/issues/51383

I like `Export()`. We could also consider `Extern()` so it is consistent with C++ and Rust.

Reply:
Added link.